R/utilities.R

```diff
@@ -2646,3 +2646,267 @@ standardize_chr_prefix = function(incoming_vector,
+    select(-total) %>%
+    `names<-`(c("sample_id", "PGA")) %>%
+    as.data.frame()
+
+  return(affected_regions)
+
+}
+
+#' This function adjusts ploidy of the sample using the percent of genome altered (PGA). The PGA is calculated internally, but can also be optionally provided as data frame
+#' if calculated from other sources. Only the samples above the threshold-provided PGA will have ploidy adjusted. The function can work with either individual or
+#' multi-sample seg file. The telomeres are always excluded from calculation, and sex chromosomes can be optionally included or excluded. The supported projections are grch37 and hg38.
+#' The chromosome prefix is handled internally per projection and does not need to be consistent.
+#'
+#' @param this_seg Input data frame of seg file.
+#' @param seg_path Optionally, specify the path to a local seg file.
+#' @param projection Argument specifying the projection of seg file, which will determine chr prefix and genome size. Default is grch37, but hg38 is also accepted.
+#' @param pga If PGA is calculated through other sources, the data frame with columns sample_id and PGA can be provided in this argument.
+#' @param pga_cutoff Minimum PGA for the sample to adjust ploidy. Default is 0.05 (5%).
+#' @param exclude_sex Boolean argument specifying whether to exclude sex chromosomes from calculation. Default is TRUE.
+#' @param return_seg Boolean argument specifying whether to return a data frame in seg-consistent format, or a raw data frame with all step-by-step transformations. Default is TRUE.
+#'
+#' @return A data frame in seg-consistent format with ploidy-adjusted log ratios.
+#' @export
+#' @import tidyverse
+#'
+#' @examples
+#' sample_seg = get_sample_cn_segments(this_sample_id = "14-36022T") %>% rename("sample"="ID")
+#' adjust_ploidy(this_seg = sample_seg)
+#'
+#' multi_sample_seg = rbind(get_sample_cn_segments(this_sample_id = "14-36022T"),
+#'                          get_sample_cn_segments(this_sample_id = "BLGSP-71-21-00243-01A-11E")) %>%
+#'                          rename("sample"="ID")
+#' adjust_ploidy(this_seg=multi_sample_seg)
+#'
+adjust_ploidy = function(this_seg,
+                         seg_path,
+                         projection = "grch37",
+                         pga,
+                         pga_cutoff = 0.05,
+                         exclude_sex = TRUE,
+                         return_seg = TRUE) {
+
+  # ensure the specified projection is correct
+  # this is only needed to adjust ploidy if the pre-adjusted ploidy is not provided
+  if (!projection %in% c("grch37", "hg38") & missing(pga)) {
+    stop(
+      "You specified projection that is currently not supported. Please provide seg files in either hg38 or grch37."
+    )
+  }
+
+  # if the seg is a local file, read it in
+  if (!missing(seg_path)) {
+    message(paste0("Reading thhe seg file from ", seg_path))
+    this_seg = read_tsv(seg_path)
+  }
+
+  # ensure consistent chromosome prefixing
```

Comment:
Does this require seg files to have specific headers? I thought some of our headers differed in the naming of the columns but I could be wrong.